server/src/main/java/org/apache/uniffle/server/storage/LocalStorageManager.java

```diff
@@ -139,32 +140,55 @@ public class LocalStorageManager extends SingleStorageManager {
 
   @Override
   public Storage selectStorage(ShuffleDataFlushEvent event) {
-    LocalStorage storage = localStorages.get(ShuffleStorageUtils.getStorageIndex(
-        localStorages.size(),
-        event.getAppId(),
-        event.getShuffleId(),
-        event.getStartPartition()));
-    if (storage.containsWriteHandler(event.getAppId(), event.getShuffleId(), event.getStartPartition())
-        && storage.isCorrupted()) {
-      LOG.error("storage " + storage.getBasePath() + " is corrupted");
-    }
-    if (storage.isCorrupted()) {
-      storage = getRepairedStorage(event.getAppId(), event.getShuffleId(), event.getStartPartition());
+    String appId = event.getAppId();
+    int shuffleId = event.getShuffleId();
+    int partitionId = event.getStartPartition();
+
+    try {
+      LocalStorage storage = partitionsOfStorage.get(appId).get(shuffleId).get(partitionId);
+      if (storage.isCorrupted()) {
+        throw new RuntimeException("LocalStorage: " + storage.getBasePath() + " is corrupted.");
+      }
+      return storage;
+    } catch (NullPointerException npe) {
+      // Ignore
     }
+
+    // Firstly getting the storage based on its (appId, shuffleId, partitionId) hash value
+    LocalStorage storage =
+        localStorages
+            .stream()
+            .filter(x -> x.canWrite() && !x.isCorrupted())
+            .collect(Collectors.toList())
+            .get(
+                ShuffleStorageUtils.getStorageIndex(
+                    localStorages.size(),
+                    appId,
+                    shuffleId,
+                    partitionId
+                )
+            );
     event.setUnderStorage(storage);
+
+    // store it to cache.
+    partitionsOfStorage.putIfAbsent(appId, Maps.newConcurrentMap());
```

Comment:
It's very convenient to use `putIfAbsent`, but `Maps.newConcurrentMap()` will be invoked every times? right?

Reply:
Nice catch. Let's use `computeIfAbsent`